### PR TITLE
base58 encoding's leading zero compression fixed and test added

### PR DESCRIPTION
**Pull request by yashbhutwala**
For now, I've added test cases in the scope of the class, but please suggest a better way if possible.

### CODE REVIEW

File: test/test_base58.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# Copyright (c) 2018 John Newbery
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test base58.py"""
+import unittest
+
+import base58
+from binascii import unhexlify
+
+TEST_CASES = [
```
Comment (jnewbery) [inline test/test_base58.py]: Great. I see that you've taken these from the bitcoind repo. Can you add a comment saying something like "test vectors from https://github.com/bitcoin/bitcoin/blob/b5e4b9b5100ec15217d43edb5f4149439f4b20a5/src/test/data/base58_encode_decode.json"

File: test/test_base58.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2018 John Newbery
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test base58.py"""
+import unittest
```
Comment (jnewbery) [inline test/test_base58.py]: Please sort according to pep8 (std library first, then third party libraries, then package modules). This would become:

```python
from binascii import unhexlify
import unittest

import base58
```

File: test/test_base58.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    ["73696d706c792061206c6f6e6720737472696e67", "2cFupjhnEsSn59qHXstmK2ffpLv2"],
+    ["00CF00816E4BE7DBC1C3DF0C2B0BE2B77E4AD99A14111E6C6F", "1KsXUqbYDGz5gvBL2mo6as5KqKueRWc2Wr"],
+    ["00000000000000000000", "1111111111"],
+    # ["  1234", ""]
+]
+
+class Base58TestCase(unittest.TestCase):
+
+    def test_bytes_to_base58(self):
+        for exp_bin, exp_base58 in TEST_CASES:
```
Comment (jnewbery) [inline test/test_base58.py]: I'm not sure what `exp_` means here. Expected? In which case, I think I'd rename as follows:

`exp_bin` -> `hex`
`exp_bin` (line below) -> `bin`

File: test/test_base58.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+from binascii import unhexlify
+
+TEST_CASES = [
+    ["", ""],
+    ["61", "2g"],
+    ["626262", "a3gV"],
+    ["73696d706c792061206c6f6e6720737472696e67", "2cFupjhnEsSn59qHXstmK2ffpLv2"],
+    ["00CF00816E4BE7DBC1C3DF0C2B0BE2B77E4AD99A14111E6C6F", "1KsXUqbYDGz5gvBL2mo6as5KqKueRWc2Wr"],
+    ["00000000000000000000", "1111111111"],
+    # ["  1234", ""]
```
Comment (jnewbery) [inline test/test_base58.py]: Remove this line?

**Review by jnewbery — COMMENTED** [review]
Thanks so much for this PR! Code change looks good. I've added a few nitty comments to the test code, but otherwise looks great.

File: test/test_base58.py (hunk 3 above)
Comment (yashbhutwala) [reply test/test_base58.py]: `hex` and `bin` are built-in function names, but I'll rename them to be `expected_hex` and `expected_bin` to be clear and descriptive.

File: test/test_base58.py (hunk 4 above)
Comment (yashbhutwala) [reply test/test_base58.py]: I can. I didn't know how the encoding should deal with unexpected edge cases.

File: test/test_base58.py (hunk 1 above)
Comment (yashbhutwala) [reply test/test_base58.py]: Sure thing!

File: test/test_base58.py (hunk 4 above)
Comment (jnewbery) [reply test/test_base58.py]: bitcoind removes leading and trailing spaces when decoding base58 strings. This library only has bytes->base58 string, so leading spaces doesn't really make sense.

File: test/test_base58.py (hunk 3 above)
Comment (jnewbery) [reply test/test_base58.py]: ah yes, of course.

Sorry, my previous comment wasn't clear. What I meant was that `exp_bin` isn't really 'expected' - it's provided. Perhaps `input_hex` and `input_bin` are better?

File: test/test_base58.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+class Base58TestCase(unittest.TestCase):
+    def test_bytes_to_base58(self):
+        for expected_hex, expected_base58 in TEST_CASES:
+            expected_bin = unhexlify(expected_hex)
+            actual_base58 = base58.bytes_to_base58(expected_bin)
+            self.assertEqual(actual_base58, expected_base58)
+
+    # TODO: test base58 decoding
+
```
Comment (jnewbery) [inline test/test_base58.py]: Remove trailing blank line.

**Comment by jnewbery** [comment]
Looks great. A couple more nitty comments.

Please squash commits to a single commit.

File: test/test_base58.py (hunk 4 above)
Comment (yashbhutwala) [reply test/test_base58.py]: Gotcha!

**Comment by jnewbery** [comment]
Thanks!
